client/file-manager/src/in_memory.rs

```diff
@@ -1,40 +1,167 @@
-use storage_hub_infra::types::{Chunk, FileProof, Key, Metadata};
+use std::collections::HashMap;
 
-use crate::traits::FileStorage;
+use common::types::HashT;
+use sp_core::H256;
+use storage_hub_infra::types::{Chunk, ChunkId, FileProof, Key, Leaf, Metadata};
 
-pub struct InMemoryFileStorage {}
+use sp_trie::{recorder::Recorder, MemoryDB, Trie, TrieDBBuilder, TrieLayout, TrieMut};
+use trie_db::TrieDBMutBuilder;
 
-impl InMemoryFileStorage {
+use crate::traits::{FileStorage, FileStorageError};
+
+pub struct FileData<T: TrieLayout + 'static> {
+    root: HashT<T>,
+    memdb: MemoryDB<T::Hash>,
+}
+
+impl<T: TrieLayout + 'static> FileData<T> {
+    pub fn get() {}
+}
+
+impl<T: TrieLayout + 'static> FileData<T> {
+    fn new() -> Self {
+        Self {
+            root: Default::default(),
+            memdb: MemoryDB::default(),
+        }
+    }
+
+    pub fn get_root(&self) -> H256 {
+        H256::from_slice(
+            self.root
+                .as_ref()
+                .try_into()
+                .expect("trie hash should be 32 bytes"),
+        )
+    }
+}
+
+pub struct InMemoryFileStorage<T: TrieLayout + 'static> {
+    pub metadata: HashMap<Key, Metadata>,
+    pub file_data: HashMap<Key, FileData<T>>,
+}
+
+impl<T: TrieLayout> InMemoryFileStorage<T> {
     pub fn new() -> Self {
-        Self {}
+        Self {
+            metadata: HashMap::new(),
+            file_data: HashMap::new(),
+        }
     }
 }
 
-impl FileStorage for InMemoryFileStorage {
-    type Key = Key;
-    type Value = Chunk;
+impl<T: TrieLayout + 'static> FileStorage for InMemoryFileStorage<T> {
+    fn generate_proof(
+        &self,
+        file_key: &Key,
+        chunk_id: &ChunkId,
+    ) -> Result<FileProof, FileStorageError> {
+        let metadata = self
+            .metadata
+            .get(file_key)
+            .ok_or(FileStorageError::FileDoesNotExist)?;
+
+        let file_data = match self.file_data.get(file_key) {
+            Some(file_data) => file_data,
+            None => panic!(
+                "Invariant broken! Metadata for file key {:?} found but no associated trie",
+                file_key
+            ),
+        };
+
+        if file_data.get_root() != metadata.fingerprint {
+            return Err(FileStorageError::FingerprintAndStoredFileMismatch);
```

Comment:
We should panic only on broken invariants that the client needs to maintain (i.e. if there is an entry in metadata, there must be an entry in the file_data, even if it does not contain any chunk yet).

In this case we check if the file is complete or not (which can happen when someone requests a read and we don't have the whole file yet).

Let me split it properly.

Reply:
Discussed in call. Result: will be split into two errors. For `debug_assert` will be used in the overlaying task to panic in case this error shows up. It is technically possible that this happens if the overlaying tasks don't handle corrupt file deletions properly. But it shouldn't happen.